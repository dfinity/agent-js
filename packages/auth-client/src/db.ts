import { openDB, IDBPDatabase } from 'idb';
import { KEY_STORAGE_DELEGATION, KEY_STORAGE_KEY } from './storage';

type Database = IDBPDatabase<unknown>;
type IDBValidKey = string | number | Date | BufferSource | IDBValidKey[];
const AUTH_DB_NAME = 'auth-client-db';
const OBJECT_STORE_NAME = 'ic-keyval';

const _openDbStore = async (
  dbName = AUTH_DB_NAME,
  storeName = OBJECT_STORE_NAME,
  version: number,
) => {
  // Clear legacy stored delegations
  if (localStorage && localStorage.getItem(KEY_STORAGE_DELEGATION)) {
    localStorage.removeItem(KEY_STORAGE_DELEGATION);
    localStorage.removeItem(KEY_STORAGE_KEY);
  }
  return await openDB(dbName, version, {
    upgrade: database => {
      database.objectStoreNames;
      if (database.objectStoreNames.contains(storeName)) {
        database.clear(storeName);
      }
      database.createObjectStore(storeName);
    },
  });
};

async function _getValue<T>(
  db: Database,
  storeName: string,
  key: IDBValidKey,
): Promise<T | undefined> {
  return await db.get(storeName, key);
}

async function _setValue<T>(
  db: Database,
  storeName: string,
  key: IDBValidKey,
  value: T,
): Promise<IDBValidKey> {
  return await db.put(storeName, value, key);
}

async function _removeValue(db: Database, storeName: string, key: IDBValidKey): Promise<void> {
  return await db.delete(storeName, key);
}

export type DBCreateOptions = {
  dbName?: string;
  storeName?: string;
  version?: number;
};

/**
 * Simple Key Value store
 * Defaults to `'auth-client-db'` with an object store of `'ic-keyval'`
 */
export class IdbKeyVal {
  /**
   *
   * @param {DBCreateOptions} options {@link DbCreateOptions}
   * @param {DBCreateOptions['dbName']} options.dbName name for the indexeddb database
   * @default 'auth-client-db'
   * @param {DBCreateOptions['storeName']} options.storeName name for the indexeddb Data Store
   * @default 'ic-keyval'
   * @param {DBCreateOptions['version']} options.version version of the database. Increment to safely upgrade
   * @constructs an {@link IdbKeyVal}
   */
  public static async create(options?: DBCreateOptions): Promise<IdbKeyVal> {
    const { dbName = AUTH_DB_NAME, storeName = OBJECT_STORE_NAME, version = 1 } = options ?? {};
    const db = await _openDbStore(dbName, storeName, version);
    return new IdbKeyVal(db, storeName);
  }

  // Do not use - instead prefer create
  private constructor(private _db: Database, private _storeName: string) {}

  /**
   * Basic setter
   * @param {IDBValidKey} key string | number | Date | BufferSource | IDBValidKey[]
   * @param value value to set
   * @returns void
   */
  public async set<T>(key: IDBValidKey, value: T) {
    return await _setValue<T>(this._db, this._storeName, key, value);
  }
  /**
   * Basic getter
   * Pass in a type T for type safety if you know the type the value will have if it is found
   * @param {IDBValidKey} key string | number | Date | BufferSource | IDBValidKey[]
   * @returns `Promise<T | null>`
   * @example
   * await get<string>('exampleKey') -> 'exampleValue'
   */
  public async get<T>(key: IDBValidKey): Promise<T | null> {
    return (await _getValue<T>(this._db, this._storeName, key)) ?? null;
  }

  /**
   * Remove a key
   * @param key {@link IDBValidKey}
   * @returns void
   */
  public async remove(key: IDBValidKey) {
    return await _removeValue(this._db, this._storeName, key);
  }
}
