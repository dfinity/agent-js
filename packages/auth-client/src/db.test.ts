import 'fake-indexeddb/auto';
import { IdbKeyVal } from './db';

let testCounter = 0;

const testDb = async () => {
  return await IdbKeyVal.create({
    dbName: 'db-' + testCounter,
    storeName: 'store-' + testCounter,
  });
};

beforeEach(() => {
  testCounter += 1;
});

describe('indexeddb wrapper', () => {
  it('should store a basic key value', async () => {
    const db = await testDb();
    const shouldSet = async () => await db.set('testKey', 'testValue');
    expect(shouldSet).not.toThrow();

    expect(await db.get('testKey')).toBe('testValue');
  });
  it('should support removing a value', async () => {
    const db = await testDb();
    await db.set('testKey', 'testValue');

    expect(await db.get('testKey')).toBe('testValue');

    await db.remove('testKey');

    expect(await db.get('testKey')).toBe(null);
  });
});
