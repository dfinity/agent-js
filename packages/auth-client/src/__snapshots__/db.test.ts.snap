// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexeddb wrapper should support storing a CryptoKeyPair 1`] = `
{
  "privateKey": {
    "algorithm": {
      "hash": {
        "name": "SHA-256",
      },
      "modulusLength": 2048,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    "extractable": true,
    "type": "private",
    "usages": [
      "decrypt",
    ],
  },
  "publicKey": {
    "algorithm": {
      "hash": {
        "name": "SHA-256",
      },
      "modulusLength": 2048,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    "extractable": true,
    "type": "public",
    "usages": [
      "encrypt",
    ],
  },
}
`;
